### PR TITLE
[SPARK-17374][SQL] Better error messages when parsing JSON using DataFrameReader

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR adds better error messages for malformed record when reading a JSON file using DataFrameReader.  

For example, for query:

```
import org.apache.spark.sql.types._
val corruptRecords = spark.sparkContext.parallelize("""{"a":{, b:3}""" :: Nil)
val schema = StructType(StructField("a", StringType, true) :: Nil)
val jsonDF = spark.read.schema(schema).json(corruptRecords)
```

**Before change:**
We silently replace corrupted line with null

```
scala> jsonDF.show
+----+
|   a|
+----+
|null|
+----+
```

**After change:**
Add an explicit warning message:

```
scala> jsonDF.show
16/09/02 14:43:16 WARN JacksonParser: Found at least one malformed records (sample: {"a":{, b:3}). The JSON reader will replace
all malformed records with placeholder null in current PERMISSIVE parser mode.
To find out which corrupted records have been replaced with null, please use the
default inferred schema instead of providing a custom schema.

Code example to print all malformed records (scala):
===================================================
// The corrupted record exists in column _corrupt_record.
val parsedJson = spark.read.json("/path/to/json/file/test.json")

+----+
|   a|
+----+
|null|
+----+
```
### 
## How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala [hunk 1]
```diff
@@ -62,8 +68,35 @@ class JacksonParser(
       throw new RuntimeException(s"Malformed line in FAILFAST mode: $record")
     }
     if (options.dropMalformed) {
-      logWarning(s"Dropping malformed line: $record")
+      if (!isWarningPrintedForMalformedRecord) {
+        logWarning(
```
Comment (clockfly) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala]: Only print a warning for the first time.

**Comment by clockfly** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala [hunk 2]
```diff
@@ -52,6 +53,11 @@ class JacksonParser(
   private val factory = new JsonFactory()
   options.setJacksonOptions(factory)
 
+  private val emptyRow: Seq[InternalRow] = Seq(new GenericInternalRow(schema.length))
+
+  @transient
+  private var isWarningPrintedForMalformedRecord: Boolean = false
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala]: private[this]?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala [hunk 3]
```diff
@@ -62,8 +68,39 @@ class JacksonParser(
...
+             |val parsedJson = spark.read.json("/path/to/json/file/test.json")
+             |
+           """.stripMargin)
+        isWarningPrintedForMalformedRecord = true
+      }
       Nil
+    } else if (schema.getFieldIndex(columnNameOfCorruptRecord).isEmpty) {
+      if (!isWarningPrintedForMalformedRecord) {
+        logWarning(
+          s"""Found at least one malformed records (sample: $record). The JSON reader will replace
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala]: should we make a common method to produce this error message?

Comment (clockfly) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala]: It is different, although similar.

**Comment by clockfly** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -1081,7 +1081,34 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
-  test("Corrupt records: PERMISSIVE mode") {
+  test("Corrupt records: PERMISSIVE mode, without designated column for malformed records") {
+    withTempView("jsonTable") {
+      val schema = StructType(
+        StructField("a", StringType, true) ::
+          StructField("b", StringType, true) ::
+          StructField("c", StringType, true) :: Nil)
+
+      val jsonDF = spark.read.schema(schema).json(corruptRecords)
+      jsonDF.createOrReplaceTempView("jsonTable")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: why create this temp view? we can
```
checkAnswer(jsonDF.select($"a", $"b", $"c"), Seq(Row...))
```

**Comment by cloud-fan** [comment]
thanks, merging to master! @clockfly can you send a follow-up PR to address the minor comment in test?

**Pull request closed without merging**